File: AOC_PY/days/day08.py
import os

INPUT_PATH = os.path.join(os.path.dirname(__file__), '../../Inputs', 'day08.txt')


class Day08:

    def __init__(self):
        with open(INPUT_PATH, 'r') as infile:
            lines = infile.read().splitlines()
        parse = lambda x: (str(x[0]), int(x[1]))
        self.program = [parse(line.split(' ')) for line in lines]

    def run_program(self, change_line=None):
        acc, line, processed = 0, 0, [False for _ in self.program]
        while 1:
            if line == len(self.program):
                return True, acc
            if processed[line]:
                return False, acc
            processed[line] = True
            op, param = self.program[line]
            if line == change_line:
                if op == 'nop':
                    op = 'jmp'
                elif op == 'jmp':
                    op = 'nop'
            if op == 'nop':
                line += 1
            elif op == 'acc':
                acc += param
                line += 1
            elif op == 'jmp':
                line += param

    def solve1(self):
        _, acc = self.run_program()
        return acc

    def solve2(self):
        for line in range(len(self.program)):
            ok, acc = self.run_program(line)
            if ok:
                return acc
        return 0


def main():
    x = Day08()
    print(x.solve1())
    print(x.solve2())


if __name__ == '__main__':
    main()

File: AOC_CS/Days/Day08.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace AOC_CS.Days
{
    class Day08
    {
        const string INPUT_FILE = "Inputs/day08.txt";

        private Tuple<string, int>[] program;

        public Day08()
        {
            program = null;
        }

        private void ReadInput()
        {
            if (program != null) return;
            var cmds = File.ReadAllLines(INPUT_FILE);
            var prg = cmds.Aggregate(
                new List<Tuple<string, int>>(),
                (acc, cmd) =>
                {
                    var tmp = cmd.Split(' ');
                    acc.Add(new Tuple<string, int>(tmp[0], int.Parse(tmp[1])));
                    return acc;
                }
            );
            program = prg.ToArray();
        }

        private bool RunProgram(int changeLine, out int accumulator)
        {
            accumulator = 0;
            var line = 0;
            var processed = new bool[program.Length];
            while (true)
            {
                if (line == program.Length) return true;
                if (processed[line]) return false;
                processed[line] = true;
                var (op, param) = program[line];
                if (line == changeLine)
                {
                    if (op == "jmp") op = "nop";
                    else if (op == "nop") op = "jmp";
                }
                switch (op)
                {
                    case "nop":
                        ++line;
                        break;
                    case "acc":
                        accumulator += param;
                        ++line;
                        break;
                    case "jmp":
                        line += param;
                        break;
                }
            }
        }

        public int Solve1()
        {
            ReadInput();
            RunProgram(-1, out var solution);
            return solution;
        }

        public int Solve2()
        {
            ReadInput();
            int solution = 0;
            for (var line = 0; line < program.Length; ++line)
            {
                if (RunProgram(line, out solution)) return solution;
            }
            return 0;
        }
    }
}
